[ONNX] New export logic leveraging ExportedProgram and ONNX IR

justinchuby opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #132530

1/n PR to

- Move code from torch-onnx from commit https://github.com/justinchuby/torch-onnx/commit/395495e566c1ee5fad438775528626cc45f88431 into torch.onnx and fixes imports.
- Integrate the new export logic with the torch.onnx.export API and include basic set of tests.
- Refactor the API for the change.
- Improve documentation.

Next PRs will be more tests and docs.

Fix https://github.com/pytorch/pytorch/issues/129277

justinchuby committed:
Update

[ghstack-poisoned]

justinchuby committed:
Update

[ghstack-poisoned]

justinchuby commented:
torch/onnx/_internal/exporter/_analysis.py:56:5 :Error Need type annotation for "parameter_count"  [MYPY/[var-annotated]]
torch/onnx/_internal/exporter/_analysis.py:57:5 :Error Need type annotation for "buffer_count"  [MYPY/[var-annotated]]
torch/onnx/_internal/exporter/_analysis.py:179:5 :Error Need type annotation for "fx_node_target_count"  [MYPY/[var-annotated]]
torch/onnx/_internal/exporter/_analysis.py:186:1 :Error Function is missing a type annotation for one or more arguments  [MYPY/[no-untyped-def]]
torch/onnx/_internal/exporter/_analysis.py:209:13 :Error Argument 1 to "from_torchlib" of "ONNXRegistry" has incompatible type "Registry"; expected "Mapping[str, OverloadedFunction] | None"  [MYPY/[arg-type]]
torch/onnx/_internal/exporter/_tensors.py:38:5 :Error Function is missing a type annotation  [MYPY/[no-untyped-def]]
torch/onnx/_internal/exporter/_tensors.py:48:5 :Error Function is missing a type annotation  [MYPY/[no-untyped-def]]
torch/onnx/_internal/exporter/_tensors.py:51:5 :Error Function is missing a type annotation  [MYPY/[no-untyped-def]]
torch/onnx/_internal/exporter/_tensors.py:54:5 :Error Function is missing a type annotation  [MYPY/[no-untyped-def]]
torch/onnx/_internal/exporter/_tensors.py:57:5 :Error Function is missing a type annotation  [MYPY/[no-untyped-def]]
torch/onnx/_internal/exporter/_tensors.py:60:5 :Error Function is missing a type annotation  [MYPY/[no-untyped-def]]
torch/onnx/_internal/exporter/_tensors.py:63:5 :Error Function is missing a type annotation  [MYPY/[no-untyped-def]]
torch/onnx/_internal/exporter/_tensors.py:66:5 :Error Function is missing a type annotation  [MYPY/[no-untyped-def]]
torch/onnx/_internal/exporter/_tensors.py:69:5 :Error Function is missing a type annotation  [MYPY/[no-untyped-def]]
torch/onnx/_internal/exporter/_tensors.py:72:5 :Error Function is missing a type annotation  [MYPY/[no-untyped-def]]
torch/onnx/_internal/exporter/_tensors.py:75:5 :Error Function is missing a type annotation  [MYPY/[no-untyped-def]]
torch/onnx/_internal/exporter/_tensors.py:78:5 :Error Function is missing a type annotation  [MYPY/[no-untyped-def]]
torch/onnx/_internal/exporter/_tensors.py:81:5 :Error Function is missing a type annotation  [MYPY/[no-untyped-def]]
torch/onnx/_internal/exporter/_tensors.py:84:5 :Error Function is missing a type annotation  [MYPY/[no-untyped-def]]
torch/onnx/_internal/exporter/_tensors.py:87:5 :Error Function is missing a type annotation  [MYPY/[no-untyped-def]]
torch/onnx/_internal/exporter/_tensors.py:90:5 :Error Function is missing a type annotation  [MYPY/[no-untyped-def]]
torch/onnx/_internal/exporter/_tensors.py:93:5 :Error Function is missing a type annotation  [MYPY/[no-untyped-def]]
torch/onnx/_internal/exporter/_tensors.py:96:5 :Error Function is missing a type annotation  [MYPY/[no-untyped-def]]
torch/onnx/_internal/exporter/_ir_passes.py:15:69 :Error Incompatible types in assignment (expression has type "str | None", target has type "str")  [MYPY/[assignment]]
torch/onnx/_internal/exporter/_ir_passes.py:21:70 :Error Incompatible types in assignment (expression has type "str | None", target has type "str")  [MYPY/[assignment]]
torch/fx/experimental/proxy_tensor.py:850:57 :Error "Callable[[], Proxy]" has no attribute "proxy"  [MYPY/[attr-defined]]
torch/nested/_internal/sdpa.py:278:14 :Error Too many arguments for "_SDPAParams"  [MYPY/[call-arg]]
torch/onnx/_internal/exporter/_isolated.py:14:1 :Error Function is missing a type annotation  [MYPY/[no-untyped-def]]
torch/onnx/_internal/exporter/_isolated.py:23:1 :Error Function is missing a return type annotation  [MYPY/[no-untyped-def]]
torch/onnx/_internal/exporter/_isolated.py:23:1 :Error Function is missing a type annotation for one or more arguments  [MYPY/[no-untyped-def]]
torch/_dynamo/compiled_autograd.py:370:13 :Error Module has no attribute "set_autograd_compiler"  [MYPY/[attr-defined]]
torch/_dynamo/compiled_autograd.py:374:9 :Error Module has no attribute "set_verbose_logger"  [MYPY/[attr-defined]]
torch/_dynamo/compiled_autograd.py:383:9 :Error Module has no attribute "set_autograd_compiler"  [MYPY/[attr-defined]]
torch/_dynamo/compiled_autograd.py:388:13 :Error Module has no attribute "set_autograd_compiler"  [MYPY/[attr-defined]]
torch/_dynamo/compiled_autograd.py:396:9 :Error Module has no attribute "set_autograd_compiler"  [MYPY/[attr-defined]]
torch/_dynamo/compiled_autograd.py:403:5 :Error Module has no attribute "set_autograd_compiler"  [MYPY/[attr-defined]]
torch/_dynamo/compiled_autograd.py:404:5 :Error Module has no attribute "set_verbose_logger"  [MYPY/[attr-defined]]
torch/backends/cuda/__init__.py:334:12 :Error Module has no attribute "_is_flash_attention_available"  [MYPY/[attr-defined]]
torch/onnx/_internal/exporter/_decomp.py:42:1 :Error Function is missing a type annotation for one or more arguments  [MYPY/[no-untyped-def]]
torch/onnx/_internal/exporter/_dispatching.py:271:146 :Advice line too long (145 > 120 characters) See https://github.com/PyCQA/flake8-bugbear#list-of-warnings [FLAKE8/B950]
torch/onnx/_internal/exporter/_analysis.py:78:160 :Advice line too long (159 > 120 characters) See https://github.com/PyCQA/flake8-bugbear#list-of-warnings [FLAKE8/B950]
torch/onnx/_internal/exporter/_analysis.py:205:9 :Advice 'onnxscript.function_libs.torch_lib.ops' imported but unused See https://www.flake8rules.com/rules/F401.html [FLAKE8/F401]
torch/serialization.py:948:5 :Error Need type annotation for "serialized_storages" (hint: "serialized_storages: dict[<type>, <type>] = ...")  [MYPY/[var-annotated]]
torch/onnx/_internal/exporter/_building.py:390: :Error This line has an unqualified `type: ignore`; please convert it to `type: ignore[xxxx]` [TYPEIGNORE/unqualified type: ignore]
torch/_subclasses/fake_tensor.py:213: :Error Name "torch._C._SchemaInfo" is not defined  [MYPY/[name-defined]]
torch/_subclasses/fake_tensor.py:214:12 :Error Module has no attribute "_SchemaInfo"  [MYPY/[attr-defined]]
torch/_subclasses/fake_tensor.py:1594:37 :Error Argument 2 to "set_" of "TensorBase" has incompatible type "int | SymInt"; expected "int"  [MYPY/[arg-type]]
torch/_subclasses/fake_tensor.py:1594:53 :Error Argument 3 to "set_" of "TensorBase" has incompatible type "tuple[int | SymInt, ...]"; expected "Size | list[int] | tuple[int, ...]"  [MYPY/[arg-type]]
torch/_subclasses/fake_tensor.py:1594:60 :Error Argument 4 to "set_" of "TensorBase" has incompatible type "tuple[int | SymInt, ...]"; expected "Size | list[int] | tuple[int, ...]"  [MYPY/[arg-type]]
torch/mtia/__init__.py:161:12 :Error Module has no attribute "_mtia_memoryStats"; maybe "_cuda_memoryStats"?  [MYPY/[attr-defined]]
torch/_dynamo/__init__.py:84:9 :Error Module has no attribute "clear_cache"  [MYPY/[attr-defined]]
torch/onnx/_internal/exporter/_capture_strategies.py:77:5 :Error Function is missing a type annotation for one or more arguments  [MYPY/[no-untyped-def]]
torch/onnx/_internal/exporter/_capture_strategies.py:100:5 :Error Function is missing a type annotation for one or more arguments  [MYPY/[no-untyped-def]]
torch/onnx/_internal/exporter/_capture_strategies.py:118:5 :Error Function is missing a type annotation for one or more arguments  [MYPY/[no-untyped-def]]
torch/onnx/_internal/exporter/_capture_strategies.py:125:5 :Error Function is missing a type annotation for one or more arguments  [MYPY/[no-untyped-def]]
torch/onnx/_internal/exporter/_capture_strategies.py:131:5 :Error Function is missing a type annotation for one or more arguments  [MYPY/[no-untyped-def]]
torch/onnx/_internal/exporter/_capture_strategies.py:137:5 :Error Function is missing a type annotation for one or more arguments  [MYPY/[no-untyped-def]]
torch/onnx/_internal/exporter/_capture_strategies.py:146:5 :Error Function is missing a type annotation for one or more arguments  [MYPY/[no-untyped-def]]
torch/onnx/_internal/exporter/_capture_strategies.py:153:5 :Error Function is missing a type annotation for one or more arguments  [MYPY/[no-untyped-def]]
torch/onnx/_internal/exporter/_capture_strategies.py:159:5 :Error Function is missing a type annotation for one or more arguments  [MYPY/[no-untyped-def]]
torch/onnx/_internal/exporter/_capture_strategies.py:165:5 :Error Function is missing a type annotation for one or more arguments  [MYPY/[no-untyped-def]]
torch/onnx/_internal/exporter/_capture_strategies.py:174:5 :Error Function is missing a type annotation for one or more arguments  [MYPY/[no-untyped-def]]
torch/onnx/_internal/exporter/_capture_strategies.py:196:5 :Error Function is missing a type annotation for one or more arguments  [MYPY/[no-untyped-def]]
torch/onnx/_internal/exporter/_capture_strategies.py:202:5 :Error Function is missing a type annotation for one or more arguments  [MYPY/[no-untyped-def]]
torch/onnx/_internal/exporter/_capture_strategies.py:208:5 :Error Function is missing a type annotation for one or more arguments  [MYPY/[no-untyped-def]]
torch/onnx/_internal/exporter/_verification.py:71:22 :Error Argument "name" to "VerificationInfo" has incompatible type "str | None"; expected "str"  [MYPY/[arg-type]]
torch/onnx/_internal/exporter/_verification.py:81:1 :Error Function is missing a return type annotation  [MYPY/[no-untyped-def]]
torch/onnx/_internal/exporter/_verification.py:84:1 :Error Cannot find implementation or library stub for module named "model_explorer"  [MYPY/[import-not-found]]
torch/onnx/_internal/exporter/_schemas.py:97: :Error This line has an unqualified `type: ignore`; please convert it to `type: ignore[xxxx]` [TYPEIGNORE/unqualified type: ignore]
torch/onnx/_internal/exporter/_schemas.py:184: :Error This line has an unqualified `type: ignore`; please convert it to `type: ignore[xxxx]` [TYPEIGNORE/unqualified type: ignore]
torch/onnx/_internal/exporter/_schemas.py:282: :Error This line has an unqualified `type: ignore`; please convert it to `type: ignore[xxxx]` [TYPEIGNORE/unqualified type: ignore]
torch/onnx/_internal/exporter/_schemas.py:318: :Error This line has an unqualified `type: ignore`; please convert it to `type: ignore[xxxx]` [TYPEIGNORE/unqualified type: ignore]
torch/onnx/_internal/exporter/_schemas.py:407: :Error This line has an unqualified `type: ignore`; please convert it to `type: ignore[xxxx]` [TYPEIGNORE/unqualified type: ignore]
torch/cuda/memory.py:69:1 :Error Module "torch._C" has no attribute "_MemPool"  [MYPY/[attr-defined]]
torch/cuda/memory.py:69:1 :Error Module "torch._C" has no attribute "_MemPoolContext"  [MYPY/[attr-defined]]
torch/onnx/_internal/exporter/_core.py:807: :Error This line has an unqualified `type: ignore`; please convert it to `type: ignore[xxxx]` [TYPEIGNORE/unqualified type: ignore]
torch/onnx/_internal/exporter/_core.py:76:159 :Advice line too long (158 > 120 characters) See https://github.com/PyCQA/flake8-bugbear#list-of-warnings [FLAKE8/B950]
torch/onnx/_internal/exporter/_core.py:78:173 :Advice line too long (172 > 120 characters) See https://github.com/PyCQA/flake8-bugbear#list-of-warnings [FLAKE8/B950]
torch/onnx/_internal/exporter/_core.py:83:135 :Advice line too long (134 > 120 characters) See https://github.com/PyCQA/flake8-bugbear#list-of-warnings [FLAKE8/B950]
torch/onnx/_internal/exporter/_core.py:86:243 :Advice line too long (242 > 120 characters) See https://github.com/PyCQA/flake8-bugbear#list-of-warnings [FLAKE8/B950]
torch/onnx/_internal/exporter/_core.py:684:9 :Advice 'onnxscript.function_libs.torch_lib.ops' imported but unused See https://www.flake8rules.com/rules/F401.html [FLAKE8/F401]
torch/onnx/_internal/exporter/_core.py:1055:13 :Advice 'onnxscript.function_libs.torch_lib.ops' imported but unused See https://www.flake8rules.com/rules/F401.html [FLAKE8/F401]
torch/onnx/_internal/fx/fx_onnx_interpreter.py:557:46 :Error "TorchScriptGraph" has no attribute "torch_graph"  [MYPY/[attr-defined]]
torch/onnx/_internal/fx/fx_onnx_interpreter.py:658:23 :Error Argument "onnx_args" to "dispatch" of "OnnxFunctionDispatcher" has incompatible type "Sequence[TorchScriptTensor | str | int | float | bool | list[Any] | complex | None]"; expected "Sequence[TensorLike | str | int | float | bool | list[Any] | complex | None]"  [MYPY/[arg-type]]
torch/onnx/_internal/fx/fx_onnx_interpreter.py:784:13 :Error Incompatible types in assignment (expression has type "TorchScriptTensor | Sequence[TorchScriptTensor]", variable has type "TorchScriptTensor | tuple[TorchScriptTensor, ...]")  [MYPY/[assignment]]
torch/fx/node.py:275:16 :Error Incompatible return value type (got "_NodeBase", expected "Node")  [MYPY/[return-value]]
torch/fx/node.py:286:16 :Error Incompatible return value type (got "_NodeBase", expected "Node")  [MYPY/[return-value]]
torch/fx/node.py:311:15 :Error "_NodeBase" has no attribute "_sort_key"  [MYPY/[attr-defined]]
torch/fx/node.py:344:9 :Error "_NodeBase" has no attribute "prepend"  [MYPY/[attr-defined]]
torch/onnx/_internal/exporter/_registration.py:84:24 :Error Incompatible types in assignment (expression has type "str", variable has type "list[str]")  [MYPY/[assignment]]
torch/onnx/_internal/exporter/_registration.py:87:24 :Error Incompatible types in assignment (expression has type "str", variable has type "list[str]")  [MYPY/[assignment]]
torch/onnx/_internal/exporter/_registration.py:96:16 :Error No overload variant of "getattr" matches argument types "Any", "list[str]"  [MYPY/[call-overload]]
torch/onnx/_internal/exporter/_registration.py:153:33 :Error Incompatible types in assignment (expression has type "Registry", variable has type "Mapping[str, OverloadedFunction] | None")  [MYPY/[assignment]]
torch/onnx/_internal/exporter/_registration.py:154:52 :Error Item "None" of "Mapping[str, OverloadedFunction] | None" has no attribute "items"  [MYPY/[union-attr]]
torch/onnx/_internal/exporter/_registration.py:212:13 :Error Name "target_or_name" already defined on line 210  [MYPY/[no-redef]]
torch/onnx/_internal/exporter/_registration.py:256:13 :Error Name "target_or_name" already defined on line 254  [MYPY/[no-redef]]
torch/onnx/_internal/exporter/_reporting.py:131:1 :Error Function is missing a return type annotation  [MYPY/[no-untyped-def]]
torch/onnx/_internal/exporter/_reporting.py:152:1 :Error Function is missing a return type annotation  [MYPY/[no-untyped-def]]
torch/onnx/_internal/exporter/_dispatching.py:192:42 :Error Item "function" of "Callable[..., Any] | str" has no attribute "_schema"  [MYPY/[union-attr]]
torch/onnx/_internal/exporter/_dispatching.py:192:42 :Error Item "str" of "Callable[..., Any] | str" has no attribute "_schema"  [MYPY/[union-attr]]
torch/onnx/_internal/exporter/_dispatching.py:218:34 :Error Item "function" of "Callable[..., Any] | str" has no attribute "_schema"  [MYPY/[union-attr]]
torch/onnx/_internal/exporter/_dispatching.py:218:34 :Error Item "str" of "Callable[..., Any] | str" has no attribute "_schema"  [MYPY/[union-attr]]
torch/onnx/_internal/exporter/_dispatching.py:256:27 :Error Incompatible types in assignment (expression has type "TypeProtocol", variable has type "tuple[Any, ...] | list[Any] | dict[str, Any] | slice | range | Node | str | int | float | bool | complex | dtype | Tensor | device | memory_format | layout | OpOverload | SymInt | SymBool | SymFloat | None")  [MYPY/[assignment]]
torch/onnx/_internal/exporter/_dispatching.py:263:27 :Error Incompatible types in assignment (expression has type "SequenceType", variable has type "tuple[Any, ...] | list[Any] | dict[str, Any] | slice | range | Node | str | int | float | bool | complex | dtype | Tensor | device | memory_format | layout | OpOverload | SymInt | SymBool | SymFloat | None")  [MYPY/[assignment]]
torch/onnx/_internal/exporter/_dispatching.py:266:27 :Error Incompatible types in assignment (expression has type "TypeProtocol", variable has type "tuple[Any, ...] | list[Any] | dict[str, Any] | slice | range | Node | str | int | float | bool | complex | dtype | Tensor | device | memory_format | layout | OpOverload | SymInt | SymBool | SymFloat | None")  [MYPY/[assignment]]
torch/onnx/_internal/exporter/_dispatching.py:274:67 :Error Argument 2 to "_attribute_type_compatible_with_arg" has incompatible type "tuple[Any, ...] | list[Any] | dict[str, Any] | slice | range | Node | str | int | float | bool | complex | dtype | Tensor | device | memory_format | layout | OpOverload | SymInt | SymBool | SymFloat | None"; expected "Value | int | float | bool | Sequence[int] | Sequence[float] | None"  [MYPY/[arg-type]]
torch/onnx/_internal/exporter/_dispatching.py:290:1 :Error Function is missing a type annotation for one or more arguments  [MYPY/[no-untyped-def]]
torch/onnx/_internal/exporter/_dispatching.py:317:41 :Error Argument 1 to "get_decomps" of "ONNXRegistry" has incompatible type "Callable[..., Any] | str"; expected "OpOverload | BuiltinFunctionType | Callable[..., Any]"  [MYPY/[arg-type]]
torch/onnx/_internal/exporter/_core.py:149:1 :Error Function is missing a return type annotation  [MYPY/[no-untyped-def]]
torch/onnx/_internal/exporter/_core.py:164:1 :Error Function is missing a return type annotation  [MYPY/[no-untyped-def]]
torch/onnx/_internal/exporter/_core.py:301:1 :Error Function is missing a return type annotation  [MYPY/[no-untyped-def]]
torch/onnx/_internal/exporter/_core.py:311:5 :Error Need type annotation for "inputs" (hint: "inputs: list[<type>] = ...")  [MYPY/[var-annotated]]
torch/onnx/_internal/exporter/_core.py:320:31 :Error Argument 1 to "append" of "list" has incompatible type "Value | Sequence[Value]"; expected "Value"  [MYPY/[arg-type]]
torch/onnx/_internal/exporter/_core.py:347:1 :Error Function is missing a return type annotation  [MYPY/[no-untyped-def]]
torch/onnx/_internal/exporter/_core.py:347:1 :Error Function is missing a type annotation for one or more arguments  [MYPY/[no-untyped-def]]
torch/onnx/_internal/exporter/_core.py:391:1 :Error Function is missing a return type annotation  [MYPY/[no-untyped-def]]
torch/onnx/_internal/exporter/_core.py:510:5 :Error Need type annotation for "constant_farm" (hint: "constant_farm: dict[<type>, <type>] = ...")  [MYPY/[var-annotated]]
torch/onnx/_internal/exporter/_core.py:558:16 :Error Incompatible return value type (got "tuple[list[tuple[Any, ...] | list[Any] | dict[str, Any] | slice | range | Node | str | int | float | complex | dtype | Tensor | device | memory_format | layout | OpOverload | SymInt | SymBool | SymFloat | None], dict[str, Any], list[str], list[str]]", expected "tuple[list[Node | None], dict[str, Any], list[str], list[str]]")  [MYPY/[return-value]]
torch/onnx/_internal/exporter/_core.py:587:47 :Error Incompatible types in assignment (expression has type "DataType", target has type "str")  [MYPY/[assignment]]
torch/onnx/_internal/exporter/_core.py:589:47 :Error Incompatible types in assignment (expression has type "tuple[Any, ...] | list[Any] | dict[str, Any] | slice | range | Node | str | int | float | complex | Tensor | memory_format | layout | OpOverload | SymInt | SymBool | SymFloat", target has type "str")  [MYPY/[assignment]]
torch/onnx/_internal/exporter/_core.py:603:24 :Error Incompatible types in assignment (expression has type "DataType", variable has type "str")  [MYPY/[assignment]]
torch/onnx/_internal/exporter/_core.py:605:24 :Error Incompatible types in assignment (expression has type "tuple[Any, ...] | list[Any] | dict[str, Any] | slice | range | Node | str | int | float | complex | Tensor | memory_format | layout | OpOverload | SymInt | SymBool | SymFloat | None", variable has type "str")  [MYPY/[assignment]]
torch/onnx/_internal/exporter/_core.py:611:12 :Error Incompatible return value type (got "tuple[list[tuple[Any, ...] | list[Any] | dict[str, Any] | slice | range | Node | str | int | float | complex | dtype | Tensor | device | memory_format | layout | OpOverload | SymInt | SymBool | SymFloat | None], dict[str, str], list[str], list[str]]", expected "tuple[list[Node | None], dict[str, Any], list[str], list[str]]")  [MYPY/[return-value]]
torch/onnx/_internal/exporter/_core.py:616:1 :Error Cannot find implementation or library stub for module named "pyinstrument"  [MYPY/[import-not-found]]
torch/onnx/_internal/exporter/_core.py:624:1 :Error Function is missing a type annotation for one or more arguments  [MYPY/[no-untyped-def]]
torch/onnx/_internal/exporter/_core.py:688:13 :Error Argument 1 to "from_torchlib" of "ONNXRegistry" has incompatible type "Registry"; expected "Mapping[str, OverloadedFunction] | None"  [MYPY/[arg-type]]
torch/onnx/_internal/exporter/_core.py:876:23 :Error Incompatible types in assignment (expression has type "Value | None", variable has type "Value")  [MYPY/[assignment]]
torch/onnx/_internal/exporter/_core.py:973:24 :Error Cannot instantiate abstract class "CaptureStrategy" with abstract attribute "_capture"  [MYPY/[abstract]]
torch/onnx/_internal/exporter/_core.py:995:12 :Error Name "result" may be undefined  [MYPY/[possibly-undefined]]
torch/onnx/_internal/exporter/_core.py:1035:12 :Error Name "program" may be undefined  [MYPY/[possibly-undefined]]
torch/onnx/_internal/exporter/_core.py:1059:17 :Error Argument 1 to "from_torchlib" of "ONNXRegistry" has incompatible type "Registry"; expected "Mapping[str, OverloadedFunction] | None"  [MYPY/[arg-type]]
torch/onnx/_internal/exporter/_core.py:1263:9 :Error Incompatible types in assignment (expression has type "VerificationInfo", variable has type "Result")  [MYPY/[assignment]]
torch/onnx/_internal/exporter/_core.py:1265:16 :Error "Result" has no attribute "absolute_difference"  [MYPY/[attr-defined]]
torch/onnx/_internal/exporter/_core.py:1268:21 :Error "Result" has no attribute "name"  [MYPY/[attr-defined]]
torch/onnx/_internal/exporter/_core.py:1269:21 :Error "Result" has no attribute "absolute_difference"  [MYPY/[attr-defined]]
torch/onnx/_internal/exporter/_core.py:1272:16 :Error "Result" has no attribute "relative_difference"  [MYPY/[attr-defined]]
torch/onnx/_internal/exporter/_core.py:1275:21 :Error "Result" has no attribute "name"  [MYPY/[attr-defined]]
torch/onnx/_internal/exporter/_core.py:1276:21 :Error "Result" has no attribute "relative_difference"  [MYPY/[attr-defined]]
torch/utils/_sympy/functions.py:9:1 :Error Module "sympy.core.numbers" has no attribute "equal_valued"  [MYPY/[attr-defined]]
torch/_inductor/dependencies.py:127:55 :Error Argument 1 to "OrderedSet" has incompatible type "set[Basic]"; expected "Iterable[Expr] | None"  [MYPY/[arg-type]]
torch/onnx/_internal/fx/serialization.py:64:30 :Error Incompatible types in assignment (expression has type "TensorProtoDataType", variable has type "int")  [MYPY/[assignment]]
torch/onnx/_internal/exporter/_schemas.py:22:5 :Error Function is missing a type annotation  [MYPY/[no-untyped-def]]
torch/onnx/_internal/exporter/_schemas.py:170:12 :Error Incompatible return value type (got "TypeProtocol", expected "TensorType | SequenceType | OptionalType")  [MYPY/[return-value]]
torch/onnx/_internal/exporter/_schemas.py:290:9 :Error Name "allowed_types" already defined on line 273  [MYPY/[no-redef]]
torch/onnx/_internal/exporter/_schemas.py:301:9 :Error Name "allowed_types" already defined on line 273  [MYPY/[no-redef]]
torch/onnx/_internal/exporter/_schemas.py:344:5 :Error Function is missing a return type annotation  [MYPY/[no-untyped-def]]
torch/onnx/_internal/exporter/_schemas.py:405:43 :Error Argument 1 to "AttributeType" has incompatible type "AttrType"; expected "int"  [MYPY/[arg-type]]
torch/onnx/_internal/exporter/_schemas.py:425:5 :Error Function is missing a type annotation for one or more arguments  [MYPY/[no-untyped-def]]
torch/onnx/_internal/exporter/_schemas.py:490:25 :Error Argument 1 to "append" of "list" has incompatible type "Parameter"; expected "AttributeParameter"  [MYPY/[arg-type]]
torch/onnx/_internal/exporter/_schemas.py:513:32 :Error Incompatible types in assignment (expression has type "list[Any]", variable has type "tuple[Any, ...]")  [MYPY/[assignment]]
torch/onnx/_internal/exporter/_building.py:79:48 :Error Incompatible types in assignment (expression has type "Value | Sequence[Value] | TensorProtocol | int | float | str | Sequence[int] | Sequence[float] | None", target has type "tuple[Value | Sequence[Value] | TensorProtocol | int | float | str | Sequence[int] | Sequence[float] | None, ...]")  [MYPY/[assignment]]
torch/onnx/_internal/exporter/_building.py:81:44 :Error Incompatible types in assignment (expression has type "Value | Sequence[Value] | TensorProtocol | int | float | bool | str | Sequence[int] | Sequence[float] | None", target has type "tuple[Value | Sequence[Value] | TensorProtocol | int | float | str | Sequence[int] | Sequence[float] | None, ...]")  [MYPY/[assignment]]
torch/onnx/_internal/exporter/_building.py:93:44 :Error Incompatible types in assignment (expression has type "None", target has type "tuple[Value | Sequence[Value] | TensorProtocol | int | float | str | Sequence[int] | Sequence[float] | None, ...]")  [MYPY/[assignment]]
torch/onnx/_internal/exporter/_building.py:102:29 :Error Incompatible types in assignment (expression has type "Value | Sequence[Value] | TensorProtocol | int | float | str | Sequence[int] | Sequence[float] | None", variable has type "TensorProtocol | int | float | bool | str | Sequence[int] | Sequence[float] | None | Attr")  [MYPY/[assignment]]
torch/onnx/_internal/exporter/_building.py:104:29 :Error Incompatible types in assignment (expression has type "Value | Sequence[Value] | TensorProtocol | int | float | bool | str | Sequence[int] | Sequence[float] | None", variable has type "TensorProtocol | int | float | bool | str | Sequence[int] | Sequence[float] | None | Attr")  [MYPY/[assignment]]
torch/onnx/_internal/exporter/_building.py:136:12 :Error Incompatible return value type (got "tuple[dict[str, tuple[Value | Sequence[Value] | TensorProtocol | int | float | str | Sequence[int] | Sequence[float] | None, ...]], dict[str, TensorProtocol | int | float | str | Sequence[int] | Sequence[float] | Attr]]", expected "tuple[dict[str, Value | Sequence[Value] | TensorProtocol | int | float | bool | str | Sequence[int] | Sequence[float] | None], dict[str, TensorProtocol | int | float | bool | str | Sequence[int] | Sequence[float] | None]]")  [MYPY/[return-value]]
torch/onnx/_internal/exporter/_building.py:288:48 :Error Argument 1 to "get" of "dict" has incompatible type "tuple[tuple[Any, ...] | Sequence[Value] | int | float | str | Sequence[int] | Sequence[float], DataType]"; expected "tuple[bool | int | float | str | TensorProtocol | tuple[int] | tuple[float], DataType]"  [MYPY/[arg-type]]
torch/onnx/_internal/exporter/_building.py:290:51 :Error Argument "value" to "tensor" has incompatible type "tuple[Any, ...] | Sequence[Value] | int | float | str | Sequence[int] | Sequence[float]"; expected "_SupportsArray[dtype[Any]] | _NestedSequence[_SupportsArray[dtype[Any]]] | bool | int | float | complex | str | bytes | _NestedSequence[bool | int | float | complex | str | bytes] | TensorProto | DLPackCompatible | ArrayCompatible"  [MYPY/[arg-type]]
torch/onnx/_internal/exporter/_building.py:292:31 :Error Invalid index type "tuple[tuple[Any, ...] | Sequence[Value] | int | float | str | Sequence[int] | Sequence[float], DataType]" for "dict[tuple[bool | int | float | str | TensorProtocol | tuple[int] | tuple[float], DataType], Value]"; expected type "tuple[bool | int | float | str | TensorProtocol | tuple[int] | tuple[float], DataType]"  [MYPY/[index]]
torch/onnx/_internal/exporter/_building.py:297:12 :Error Incompatible return value type (got "dict[str, Value | Sequence[Value] | TensorProtocol | int | float | bool | str | Sequence[int] | Sequence[float] | None]", expected "dict[str, Value | None]")  [MYPY/[return-value]]
torch/onnx/_internal/exporter/_building.py:331:12 :Error Item "RefAttr" of "Attr | RefAttr" has no attribute "value"  [MYPY/[union-attr]]
torch/onnx/_internal/exporter/_building.py:349:9 :Error Need type annotation for "nodes" (hint: "nodes: list[<type>] = ...")  [MYPY/[var-annotated]]
torch/onnx/_internal/exporter/_building.py:395:9 :Error Argument 2 of "eval" is incompatible with supertype "Evaluator"; supertype defines the argument type as "Sequence[ExtendedModeValue]"  [MYPY/[override]]
torch/onnx/_internal/exporter/_building.py:417:24 :Error Item "Sequence[Value]" of "Value | Sequence[Value] | TensorProtocol | int | float | bool | str | Sequence[int] | Sequence[float] | None" has no attribute "dtype"  [MYPY/[union-attr]]
torch/onnx/_internal/exporter/_building.py:417:24 :Error Item "int" of "Value | Sequence[Value] | TensorProtocol | int | float | bool | str | Sequence[int] | Sequence[float] | None" has no attribute "dtype"  [MYPY/[union-attr]]
torch/onnx/_internal/exporter/_building.py:417:24 :Error Item "float" of "Value | Sequence[Value] | TensorProtocol | int | float | bool | str | Sequence[int] | Sequence[float] | None" has no attribute "dtype"  [MYPY/[union-attr]]
torch/onnx/_internal/exporter/_building.py:417:24 :Error Item "bool" of "Value | Sequence[Value] | TensorProtocol | int | float | bool | str | Sequence[int] | Sequence[float] | None" has no attribute "dtype"  [MYPY/[union-attr]]
torch/onnx/_internal/exporter/_building.py:417:24 :Error Item "str" of "Value | Sequence[Value] | TensorProtocol | int | float | bool | str | Sequence[int] | Sequence[float] | None" has no attribute "dtype"  [MYPY/[union-attr]]
torch/onnx/_internal/exporter/_building.py:417:24 :Error Item "Sequence[int]" of "Value | Sequence[Value] | TensorProtocol | int | float | bool | str | Sequence[int] | Sequence[float] | None" has no attribute "dtype"  [MYPY/[union-attr]]
torch/onnx/_internal/exporter/_building.py:417:24 :Error Item "Sequence[float]" of "Value | Sequence[Value] | TensorProtocol | int | float | bool | str | Sequence[int] | Sequence[float] | None" has no attribute "dtype"  [MYPY/[union-attr]]
torch/onnx/_internal/exporter/_building.py:417:24 :Error Item "object" of "Value | Sequence[Value] | TensorProtocol | int | float | bool | str | Sequence[int] | Sequence[float] | None" has no attribute "dtype"  [MYPY/[union-attr]]
torch/onnx/_internal/exporter/_building.py:419:32 :Error Incompatible return value type (got "Value | Sequence[Value] | TensorProtocol | int | float | bool | str | Sequence[int] | Sequence[float] | None", expected "SymbolicTensor | Sequence[SymbolicTensor]")  [MYPY/[return-value]]
torch/onnx/_internal/exporter/_building.py:422:62 :Error Item "Sequence[Value]" of "Value | Sequence[Value] | TensorProtocol | int | float | bool | str | Sequence[int] | Sequence[float] | None" has no attribute "dtype"  [MYPY/[union-attr]]
torch/onnx/_internal/exporter/_building.py:422:62 :Error Item "int" of "Value | Sequence[Value] | TensorProtocol | int | float | bool | str | Sequence[int] | Sequence[float] | None" has no attribute "dtype"  [MYPY/[union-attr]]
torch/onnx/_internal/exporter/_building.py:422:62 :Error Item "float" of "Value | Sequence[Value] | TensorProtocol | int | float | bool | str | Sequence[int] | Sequence[float] | None" has no attribute "dtype"  [MYPY/[union-attr]]
torch/onnx/_internal/exporter/_building.py:422:62 :Error Item "bool" of "Value | Sequence[Value] | TensorProtocol | int | float | bool | str | Sequence[int] | Sequence[float] | None" has no attribute "dtype"  [MYPY/[union-attr]]
torch/onnx/_internal/exporter/_building.py:422:62 :Error Item "str" of "Value | Sequence[Value] | TensorProtocol | int | float | bool | str | Sequence[int] | Sequence[float] | None" has no attribute "dtype"  [MYPY/[union-attr]]
torch/onnx/_internal/exporter/_building.py:422:62 :Error Item "Sequence[int]" of "Value | Sequence[Value] | TensorProtocol | int | float | bool | str | Sequence[int] | Sequence[float] | None" has no attribute "dtype"  [MYPY/[union-attr]]
torch/onnx/_internal/exporter/_building.py:422:62 :Error Item "Sequence[float]" of "Value | Sequence[Value] | TensorProtocol | int | float | bool | str | Sequence[int] | Sequence[float] | None" has no attribute "dtype"  [MYPY/[union-attr]]
torch/onnx/_internal/exporter/_building.py:422:62 :Error Item "object" of "Value | Sequence[Value] | TensorProtocol | int | float | bool | str | Sequence[int] | Sequence[float] | None" has no attribute "dtype"  [MYPY/[union-attr]]
torch/onnx/_internal/exporter/_onnx_program.py:23:1 :Error Skipping analyzing "onnxruntime": module is installed, but missing library stubs or py.typed marker  [MYPY/[import-untyped]]
torch/onnx/_internal/exporter/_onnx_program.py:67:5 :Error Function is missing a type annotation for one or more arguments  [MYPY/[no-untyped-def]]
torch/onnx/_internal/exporter/_onnx_program.py:96:5 :Error Function is missing a return type annotation  [MYPY/[no-untyped-def]]
torch/onnx/_internal/exporter/_onnx_program.py:96:5 :Error Function is missing a type annotation for one or more arguments  [MYPY/[no-untyped-def]]
torch/onnx/_internal/exporter/_onnx_program.py:188:21 :Error Incompatible types in assignment (expression has type "bytes", variable has type "str")  [MYPY/[assignment]]
torch/onnx/_internal/exporter/_onnx_program.py:207:1 :Error Function is missing a type annotation for one or more arguments  [MYPY/[no-untyped-def]]
torch/onnx/_internal/exporter/_onnx_program.py:216:1 :Error Function is missing a type annotation  [MYPY/[no-untyped-def]]
torch/onnx/_internal/exporter/_onnx_program.py:221:1 :Error Function is missing a type annotation  [MYPY/[no-untyped-def]]
torch/onnx/_internal/exporter/_onnx_program.py:225:1 :Error Function is missing a type annotation  [MYPY/[no-untyped-def]]
torch/onnx/_internal/exporter/_onnx_program.py:266:1 :Error Function is missing a type annotation  [MYPY/[no-untyped-def]]

titaiwangms commented on test/test_public_bindings.py:
```diff
@@ -288,6 +288,21 @@ def onerror(modname):
+            "torch.onnx._internal.exporter._reporting"
+            "torch.onnx._internal.exporter._schemas"
+            "torch.onnx._internal.exporter._tensors"
+            "torch.onnx._internal.exporter._verification"
```
Looks like this is happening again...

justinchuby replied:
This is needed and it is internal so I think we should be fine

justinchuby committed:
Update

[ghstack-poisoned]

justinchuby committed:
Update

[ghstack-poisoned]

justinchuby committed:
Update

[ghstack-poisoned]

justinchuby committed:
Update

[ghstack-poisoned]

justinchuby committed:
Update

[ghstack-poisoned]

justinchuby commented on test/test_public_bindings.py:
```diff
@@ -286,6 +286,22 @@ def onerror(modname):
         # do not get imported by public code.
         private_allowlist = {
             "torch._inductor.codegen.cuda.cuda_kernel",
+            "torch.onnx._internal.exporter",
```
Add a todo

justinchuby replied:
Done

justinchuby committed:
Update

[ghstack-poisoned]

justinchuby committed:
Update

[ghstack-poisoned]

justinchuby committed:
Update

[ghstack-poisoned]

justinchuby committed:
Update

[ghstack-poisoned]

justinchuby committed:
Update

[ghstack-poisoned]

justinchuby committed:
Update on "[ONNX] New export logic leveraging ExportedProgram and ONNX IR"


Move code from torch-onnx from commit https://github.com/justinchuby/torch-onnx/commit/395495e566c1ee5fad438775528626cc45f88431 into torch.onnx and fixes imports. No tests, no usage, just source. Integration will be in following PRs (need to resolve imports and expose apis etc.).

[ghstack-poisoned]

justinchuby committed:
Update on "[ONNX] New export logic leveraging ExportedProgram and ONNX IR"


Move code from torch-onnx from commit https://github.com/justinchuby/torch-onnx/commit/395495e566c1ee5fad438775528626cc45f88431 into torch.onnx and fixes imports. No tests, no usage, just source. Integration will be in following PRs (need to resolve imports and expose apis etc.).

[ghstack-poisoned]

justinchuby committed:
Update on "[ONNX] New export logic leveraging ExportedProgram and ONNX IR"


Move code from torch-onnx from commit https://github.com/justinchuby/torch-onnx/commit/395495e566c1ee5fad438775528626cc45f88431 into torch.onnx and fixes imports. No tests, no usage, just source. Integration will be in following PRs (need to resolve imports and expose apis etc.).

[ghstack-poisoned]

justinchuby commented:
@pytorchbot merge -i

justinchuby committed:
Update on "[ONNX] New export logic leveraging ExportedProgram and ONNX IR"


1/n PR to

- Move code from torch-onnx from commit https://github.com/justinchuby/torch-onnx/commit/395495e566c1ee5fad438775528626cc45f88431 into torch.onnx and fixes imports.
- Integrate the new export logic with the torch.onnx.export API and include basic set of tests.
- Refactor the API for the change.
- Improve documentation.

Next PRs will be more tests and docs.

Fix https://github.com/pytorch/pytorch/issues/129277

[ghstack-poisoned]

justinchuby committed:
Update on "[ONNX] New export logic leveraging ExportedProgram and ONNX IR"


1/n PR to

- Move code from torch-onnx from commit https://github.com/justinchuby/torch-onnx/commit/395495e566c1ee5fad438775528626cc45f88431 into torch.onnx and fixes imports.
- Integrate the new export logic with the torch.onnx.export API and include basic set of tests.
- Refactor the API for the change.
- Improve documentation.

Next PRs will be more tests and docs.

Fix https://github.com/pytorch/pytorch/issues/129277

[ghstack-poisoned]

justinchuby commented:
@pytorchbot merge -i

justinchuby commented:
@pytorchbot merge -i

justinchuby commented:
@pytorchbot merge -i

ZainRizvi commented:
@pytorchbot revert -c nosignal -m "Sorry but it seems like Dr. CI incorrectly flagged the [pull / linux-docs / build-docs-python-false](https://hud.pytorch.org/pr/pytorch/pytorch/132530#28918577682) failure as being flaky. The job started failing consistently on CI once your PR was merged. [GH job link](https://github.com/pytorch/pytorch/actions/runs/10454830880/job/28949386844) [HUD commit link](https://hud.pytorch.org/pytorch/pytorch/commit/5fab35d77c7d1db7dbb9d5c516254a510b4f4f64)"

justinchuby commented:
@ZainRizvi Thanks, the docs error seems to be

```
E0819 14:25:38.124821 144 torch/_guards.py:281] [17/0] Error while creating guard:
E0819 14:25:38.124821 144 torch/_guards.py:281] [17/0] Name: ''
E0819 14:25:38.124821 144 torch/_guards.py:281] [17/0]     Source: shape_env
E0819 14:25:38.124821 144 torch/_guards.py:281] [17/0]     Create Function: SHAPE_ENV
E0819 14:25:38.124821 144 torch/_guards.py:281] [17/0]     Guard Types: None
E0819 14:25:38.124821 144 torch/_guards.py:281] [17/0]     Code List: None
E0819 14:25:38.124821 144 torch/_guards.py:281] [17/0]     Object Weakref: None
E0819 14:25:38.124821 144 torch/_guards.py:281] [17/0]     Guarded Class Weakref: None
E0819 14:25:38.124821 144 torch/_guards.py:281] [17/0] Traceback (most recent call last):
E0819 14:25:38.124821 144 torch/_guards.py:281] [17/0]   File "/opt/conda/envs/py_3.8/lib/python3.8/site-packages/torch/_guards.py", line 279, in create
E0819 14:25:38.124821 144 torch/_guards.py:281] [17/0]     return self.create_fn(builder, self)
E0819 14:25:38.124821 144 torch/_guards.py:281] [17/0]   File "/opt/conda/envs/py_3.8/lib/python3.8/site-packages/torch/_dynamo/guards.py", line 1780, in SHAPE_ENV
E0819 14:25:38.124821 144 torch/_guards.py:281] [17/0]     guards = output_graph.shape_env.produce_guards(
E0819 14:25:38.124821 144 torch/_guards.py:281] [17/0]   File "/opt/conda/envs/py_3.8/lib/python3.8/site-packages/torch/fx/experimental/symbolic_shapes.py", line 4185, in produce_guards
E0819 14:25:38.124821 144 torch/_guards.py:281] [17/0]     raise ConstraintViolationError(
E0819 14:25:38.124821 144 torch/_guards.py:281] [17/0] torch.fx.experimental.symbolic_shapes.ConstraintViolationError: Constraints violated (dim0_x)! For more information, run with TORCH_LOGS="+dynamic".
E0819 14:25:38.124821 144 torch/_guards.py:281] [17/0]   - Not all values of dim0_x = L['x'].size()[0] in the specified range satisfy the generated guard RoundToInt(IntTrueDiv(L['x'].size()[0], 2)) <= L['x'].size()[0].
E0819 14:25:38.125489 144 torch/_guards.py:283] [17/0] Created at:
E0819 14:25:38.125489 144 torch/_guards.py:283] [17/0]   File "/opt/conda/envs/py_3.8/lib/python3.8/site-packages/torch/_dynamo/convert_frame.py", line 603, in transform
E0819 14:25:38.125489 144 torch/_guards.py:283] [17/0]     tracer = InstructionTranslator(
E0819 14:25:38.125489 144 torch/_guards.py:283] [17/0]   File "/opt/conda/envs/py_3.8/lib/python3.8/site-packages/torch/_dynamo/symbolic_convert.py", line 2631, in __init__
E0819 14:25:38.125489 144 torch/_guards.py:283] [17/0]     output=OutputGraph(
E0819 14:25:38.125489 144 torch/_guards.py:283] [17/0]   File "/opt/conda/envs/py_3.8/lib/python3.8/site-packages/torch/_dynamo/output_graph.py", line 316, in __init__
E0819 14:25:38.125489 144 torch/_guards.py:283] [17/0]     self.init_ambient_guards()
E0819 14:25:38.125489 144 torch/_guards.py:283] [17/0]   File "/opt/conda/envs/py_3.8/lib/python3.8/site-packages/torch/_dynamo/output_graph.py", line 455, in init_ambient_guards
E0819 14:25:38.125489 144 torch/_guards.py:283] [17/0]     self.guards.add(ShapeEnvSource().make_guard(GuardBuilder.SHAPE_ENV))
```

Do you have an idea on what may have caused it? This change doesn't seem to add any new docs or introduce changes that will trigger this

justinchuby commented:
I think I know what is going on. I am going to rebase after #133825

justinchuby commented:
@pytorchbot rebase

justinchuby committed:
Update on "[ONNX] New export logic leveraging ExportedProgram and ONNX IR"


1/n PR to

- Move code from torch-onnx from commit https://github.com/justinchuby/torch-onnx/commit/395495e566c1ee5fad438775528626cc45f88431 into torch.onnx and fixes imports.
- Integrate the new export logic with the torch.onnx.export API and include basic set of tests.
- Refactor the API for the change.
- Improve documentation.

Next PRs will be more tests and docs.

Fix https://github.com/pytorch/pytorch/issues/129277

[ghstack-poisoned]

justinchuby committed:
Update on "[ONNX] New export logic leveraging ExportedProgram and ONNX IR"


1/n PR to

- Move code from torch-onnx from commit https://github.com/justinchuby/torch-onnx/commit/395495e566c1ee5fad438775528626cc45f88431 into torch.onnx and fixes imports.
- Integrate the new export logic with the torch.onnx.export API and include basic set of tests.
- Refactor the API for the change.
- Improve documentation.

Next PRs will be more tests and docs.

Fix https://github.com/pytorch/pytorch/issues/129277

[ghstack-poisoned]

justinchuby commented:
@pytorchbot merge -f "Rocm tests unrelated."

titaiwangms commented on torch/onnx/_internal/exporter/_building.py:
```diff
@@ -0,0 +1,516 @@
+                arg = tuple(arg)  # noqa: PLW2901
+            constant_value = constant_farm.get((arg, dtype))  # type: ignore[arg-type]
+            if constant_value is None:
+                constant_tensor = ir.tensor(value=arg, dtype=dtype)  # type: ignore[arg-type]
```
Is this a typo? it.Tensor?

justinchuby replied:
This is expected. ir.tensor will create a tensor.
